### PR TITLE
refactor: Allow track selection without reference surface

### PR DESCRIPTION
**Pull request by benjaminhuth**
Adds a flag to allow track selection without reference surface, which is off by default. Also refactors the track selector config handling in python to avoid that different uses get out of sync.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new configuration option, `requireReferenceSurface`, for track validation, enhancing track selection flexibility.
	- Added a function, `trackSelectorDefaultKWArgs`, to streamline track selection parameter configuration.

- **Improvements**
	- Refactored track validation logic for improved clarity and separation of concerns.
	- Enhanced configurability of track selection and seeding algorithms, allowing for more adaptable configurations.
	- Updated default values in the `TrackSelectorConfig` to accommodate the new field.

### CODE REVIEW

File: Core/include/Acts/TrackFinding/TrackSelector.hpp [hunk 1]
```diff
@@ -80,6 +80,10 @@ class TrackSelector {
     std::size_t maxSharedHits = std::numeric_limits<std::size_t>::max();
     double maxChi2 = inf;
 
+    // Whether a reference surface is required for the track
+    // If false, the parameter cuts are not evaluated
```
Comment (andiwand) [inline Core/include/Acts/TrackFinding/TrackSelector.hpp:84]:
```suggestion
    /// Whether a reference surface is required for the track
    /// If false, the parameter cuts are not evaluated
```

File: Examples/Python/python/acts/examples/reconstruction.py [hunk 2]
```diff
@@ -130,10 +130,43 @@
...
         "nMeasurementsGroupMin",
+        "requireReferenceSurface",
     ],
-    defaults=[(None, None)] * 7 + [None] * 7,
+    defaults=[(None, None)] * 7 + [None] * 8,
 )
 
+
+# Encapsulate this boilerplate code into a function so different uses do not get out of sync
+# Gets an additional lambda for the etaMax, because we might want to use this with eta binning
```
Comment (andiwand) [inline Examples/Python/python/acts/examples/reconstruction.py:140]: I believe the doc needs to be in the function?

File: Examples/Python/python/acts/examples/reconstruction.py [hunk 3]
```diff
@@ -130,10 +130,43 @@
...
+def convertPyTrackSelectorConfig(pyCfg, etaMaxLambda=lambda etaMax: etaMax):
```
Comment (andiwand) [inline Examples/Python/python/acts/examples/reconstruction.py:141]:
```suggestion
def convertPyTrackSelectorConfig(c, etaMaxLambda=lambda etaMax: etaMax):
```

File: Examples/Python/python/acts/examples/reconstruction.py [hunk 4]
```diff
@@ -130,10 +130,43 @@
...
+            loc1Min=c.loc1[0],
+            loc1Max=c.loc1[1],
+            timeMin=c.time[0],
+            timeMax=c.time[1],
+            phiMin=c.phi[0],
+            phiMax=c.phi[1],
+            etaMin=c.eta[0],
+            etaMax=c.eta[1],
+            absEtaMin=c.absEta[0],
+            absEtaMax=etaMaxLambda(c.absEta[1]),
```
Comment (andiwand) [inline Examples/Python/python/acts/examples/reconstruction.py:155]:
```suggestion
            absEtaMax= absEtaMaxLambda(c.absEta[1]),
```

File: Examples/Python/python/acts/examples/reconstruction.py [hunk 5]
```diff
@@ -1472,30 +1505,8 @@ def addCKFTracks(
...
-                minMeasurements=c.nMeasurementsMin,
-                maxHoles=c.maxHoles,
-                maxOutliers=c.maxOutliers,
-                maxHolesAndOutliers=c.maxHolesAndOutliers,
-                maxSharedHits=c.maxSharedHits,
-                maxChi2=c.maxChi2,
-                measurementCounter=c.nMeasurementsGroupMin,
-            )
+        convertPyTrackSelectorConfig(
+            c, lambda etaMax: etaMax if len(tslist) == 1 else None
```
Comment (andiwand) [inline Examples/Python/python/acts/examples/reconstruction.py:1509]: alternatively we could convert he python named tuple to a kwargs so we can still overwrite `absEtaMax` later. the lambda looks a bit arbitrary and hacky

File: Examples/Python/python/acts/examples/reconstruction.py (hunk 2 above)
Comment (benjaminhuth) [reply Examples/Python/python/acts/examples/reconstruction.py:140]: true, don't know why I did it this way^^

File: Examples/Python/python/acts/examples/reconstruction.py (hunk 5 above)
Comment (benjaminhuth) [reply Examples/Python/python/acts/examples/reconstruction.py:1509]: yeah I wanted to do this with least possible changes, and also didn't spend a lot time into thinking about this... But I don't see a way to do this within the list comprehension actually... would you do another loop?

File: Examples/Python/python/acts/examples/reconstruction.py [hunk 6]
```diff
@@ -123,10 +123,44 @@
...
+    defaults=[(None, None)] * 7 + [None] * 8,
 )
 
+
+def trackSelectorDefaultKWArgs(c):
+    """
+    Encapsulate this boilerplate code into a function so different uses do not get out of sync
+    """
+    return {
+        **acts.examples.defaultKWArgs(
```
Comment (andiwand) [inline Examples/Python/python/acts/examples/reconstruction.py:137]: nitpick: I think this should give the same?
```suggestion
    return acts.examples.defaultKWArgs(
```

**Review by andiwand — DISMISSED** [review]
LGTM

**Review by andiwand — DISMISSED** [review]

**Comment by paulgessinger** [comment]
I think there are some CI failures still.

**Comment by benjaminhuth** [comment]
I'm aware, I hope I find some time later today

**Review by AJPfleger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [9173bb4b2c99ea806eb5ad342af9d64d3ee32e1f]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608503)||
|:green_circle: | [run_art_test: test_run4_acts_vertex_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608502)||
|:green_circle: | [run_art_test: test_run4_ttbar_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608501)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608500)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608499)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608498)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608497)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608496)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608495)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608494)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608493)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflowHgtd](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608492)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608491)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608490)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608489)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608488)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608487)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608486)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608485)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608484)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608483)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608482)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608481)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608480)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608479)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608478)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608477)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608476)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608475)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608474)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608473)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608472)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608471)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608470)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608469)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46608468)||

**Pull request merged**